addons/binding/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/handler/LGWebOSHandler.java

```diff
@@ -0,0 +1,223 @@
+        } else {
+            device.addListener(this);
+            if (device.isConnected()) {
+                updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "Connected");
+            } else {
+                device.connect();
+            }
+        }
+    }
+
+    @Override
+    public void dispose() {
+        super.dispose();
+        ConnectableDevice device = getDevice();
+        if (device != null) {
+            device.removeListener(this);
+        }
+        this.discoveryManager.removeListener(this);
+        this.discoveryManager = null;
+    }
+    // Connectable Device Listener
+
+    @Override
+    public void onDeviceReady(ConnectableDevice device) { // this gets called on connection success
+        updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "Connected");
+        refreshAllChannelSubscriptions(device);
+        channelHandlers.entrySet().forEach(e -> e.getValue().onDeviceReady(device, e.getKey(), this));
+    }
+
+    @Override
+    public void onDeviceDisconnected(ConnectableDevice device) {
+        logger.debug("Device disconnected: {}", device);
+        for (Map.Entry<String, ChannelHandler> e : channelHandlers.entrySet()) {
+            e.getValue().onDeviceRemoved(device, e.getKey(), this);
+            e.getValue().removeAnySubscription(device);
+        }
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE, "TV is off");
+    }
+
+    @Override
+    public void onPairingRequired(ConnectableDevice device, DeviceService service, PairingType pairingType) {
+        updateStatus(this.thing.getStatus(), ThingStatusDetail.CONFIGURATION_PENDING, "Pairing Required");
+    }
+
+    @Override
+    public void onCapabilityUpdated(ConnectableDevice device, List<String> added, List<String> removed) {
+        logger.debug("Capabilities updated: {} - added: {} - removed: {}", device, added, removed);
+        refreshAllChannelSubscriptions(device);
+    }
+
+    @Override
+    public void onConnectionFailed(ConnectableDevice device, ServiceCommandError error) {
+        logger.debug("Connection failed: {} - error: {}", device, error.getMessage());
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Connection Failed");
+    }
+
+    // callback methods for commandHandlers
+    public void postUpdate(String channelUID, State state) {
+        this.updateState(channelUID, state);
+    }
+
+    public boolean isChannelInUse(String channelId) {
+        return this.isLinked(channelId);
+    }
+
+    // channel linking modifications
+
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
+        refreshChannelSubscription(channelUID);
+    }
+
+    @Override
+    public void channelUnlinked(ChannelUID channelUID) {
+        refreshChannelSubscription(channelUID);
+    }
+
+    // private helpers
+
+    /**
+     * Refresh channel subscription for one specific channel.
+     *
+     * @param channelUID must not be <code>null</code>
```

Comment:
Where can I read about this? not familiar with `@NonNullByDefault`

Reply:
The has recently been added a [guideline](https://github.com/triller-telekom/smarthome/blob/e4cb58328c62c74b1f4df9fca17bf9f8f4bd8110/docs/documentation/development/guidelines.md) to the smarthome framework - point A.12.
OH bindings can (and maybe will have to, soon) follow these guidelines. IDE will then detect incorrect, redundant and missing handling of nulls. There is however an issue with calls to the libraries, which are not aware of the annotations, like jdk. These places can have warning suppressed. Would be interesting to see if enabling this on your classes shows benefits or is this inconvenience.

Reply:
I read the guideline. Could you recommend an example binding to look at?

Reply:
There seem to be no bindings that currently implement that. These annotations are gradually introduced to the smarthome framework, so you may just search for NonNullByDefault in your workspace and find examples. Another example, though not finished yet, is my attempt to bring them into my binding: #2604. 
There is also an ongoing [discussion about observations in the smarthome project](https://github.com/eclipse/smarthome/issues/4321).

I think it may make no point to delay merge of this PR and wait for the null annotations, but I would like to welcome you to the discussions and share your observations.